[WIP] Service Dialogs form conversion from Angular to React

elsamaryv opened the pull request:
- Revamp of service dialogs (Also includes its conversion from Angular to React)

@miq-bot add-label enhancement
@miq-bot add-label wip

<!--
2. If this fixes an existing issue, please specify in `Fixes #<id>` format
   (As described in https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue)
-->

<!--
3. Ask @miq-bot to apply a scope label (bug, enhancement, etc) and any additional reviewers or assignees.
   (As described in https://github.com/ManageIQ/miq_bot#requested-tasks)
   e.g. `@miq-bot add-label label_name`
        `@miq-bot add-reviewer @name`
-->

elsamaryv committed:
Converts service dialogs from angular to react

elsamaryv committed:
Removes comments

elsamaryv committed:
Removes commented(unwanted) code

Fryguy commented on cypress/support/commands/common.js:
```diff
@@ -0,0 +1,37 @@
+/* eslint-disable no-undef */
+
+// Look for notification popups and disable them if present
```
Are these for "real" notifications or for debug-type notifications?  If the latter, then this shouldn't be needed because if you run the webpack-dev-server and the rails server with `CYPRESS=true` then those debug-type notifications won't be there.

Fryguy commented on cypress/support/commands/customization/service-dialogs.js:
```diff
@@ -0,0 +1,182 @@
+});
+
+// Login and navigate to add a new service dialog
+Cypress.Commands.add('navigateToAddDialog', () => {
```
well, maybe not this one because it also includes login and navigation

Fryguy commented on app/javascript/packs/component-definitions-common.js:
```diff
@@ -369,3 +372,4 @@ ManageIQ.component.addReact('WorkflowRepositoryForm', WorkflowRepositoryForm);
 ManageIQ.component.addReact('XmlHolder', XmlHolder);
 ManageIQ.component.addReact('ZoneForm', ZoneForm);
 ManageIQ.component.addReact('MiqAeClass', MiqAeClass);
+ManageIQ.component.addReact('CustomDateTimePicker', CustomDateTimePicker);
```
Same on sorting.

elsamaryv committed:
Fix issue with embedded entrypoint modal

elsamaryv committed:
Fixes issue with conditional rendering of entrypoint fields

elsamaryv committed:
Updates for entry_point_workflow field

asirvadAbrahamVarghese commented on app/javascript/components/common/inline-flash-message/index.jsx:
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
```
I think in React-17+, JSX works without importing React, so we should be able to drop it post our upgrade 👍

asirvadAbrahamVarghese commented on app/javascript/components/common/inline-flash-message/index.jsx:
```diff
@@ -0,0 +1,36 @@
+import PropTypes from 'prop-types';
+import { InlineNotification } from 'carbon-components-react';
+
+const InlineFlashMessage = ({ message, setMessage }) => {
```
- if it’s just to pass through any actions required for the close button click  maybe we could rename the prop to something like `onCloseClick` (I’m clueless when it comes to naming, so it’s your call)
 - Should we consider changing `message` to `messageInfo`, given that its an object? - again thats your call

asirvadAbrahamVarghese commented on app/javascript/components/common/inline-flash-message/index.jsx:
```diff
@@ -0,0 +1,36 @@
+      title={message.title || ''}
+      subtitle={message.subtitle || ''}
+      lowContrast
+      hideCloseButton={!setMessage}
```
It might be better to control `hideCloseButton` through a separate prop and set a default for it 🤔

asirvadAbrahamVarghese commented on app/javascript/components/common/inline-flash-message/index.jsx:
```diff
@@ -0,0 +1,36 @@
+      lowContrast
+      hideCloseButton={!setMessage}
+      onCloseButtonClick={
+        typeof setMessage === 'function' ? () => setMessage(null) : undefined
```
I was thinking if we should directly do:
`onCloseButtonClick={setMessage}`
So that react will pass down the event automatically when setMessage handler is called similar to how we’d pass the event manually like:
`onCloseButtonClick={(event)=>setMessage(event)}`

So if in case we want to access any event methods (preventDefault, stopPropagation) from this component sometime later we can do:
`<InlineFlashMessage setMessage={(event)=>{ event.preventDefault(); }} .... />`

elsamaryv committed:
Fixes sorting and adds description for component props

elsamaryv committed:
Updates InlineFlashMessage component

asirvadAbrahamVarghese commented on app/javascript/components/date-time-picker/index.jsx:
```diff
@@ -0,0 +1,95 @@
+} from 'carbon-components-react';
+import { getCurrentDate, getCurrentTimeAndPeriod } from '../service-dialog-form/helper';
+
+const CustomDateTimePicker = (field) => {
```
Correct me here if I am wrong
We will be using the component like:
```
        <CustomDateTimePicker
          label="Custom Date Time Picker"
          initialData={{ date: '11/06/2025', time: '23:11', period: 'PM' }}
          onChange={(data) => {
            // any actions with on change data
          }}
        />
```
So I was thinking - just like we did with InlineFlashMessage, inline destructuring of component props is generally cleaner instead of accessing them via `field` object:
`const CustomDateTimePicker = ({ label, onChange, initialData }) => {...`

and then add props validations and default props for them:
```
CustomDateTimePicker.propTypes = {
  initialData: PropTypes.shape({
    date: PropTypes.string,
    time: PropTypes.string,
    period: PropTypes.string,
  }),
  onChange: PropTypes.func,
  label: PropTypes.string,
};

CustomDateTimePicker.defaultProps = {
  initialData: { date: '', time: '', period: '' },
  onChange: () => {},
  label: '',
};
```
Note: date, time and period having an empty string as default value (`{ date: '', time: '', period: '' }`) shouldn't be a problem as they are falsy in JS, fallback methods like `getCurrentDate` and `getCurrentTimeAndPeriod` will still be triggered to set states initial values

asirvadAbrahamVarghese commented on app/javascript/components/date-time-picker/index.jsx:
```diff
@@ -0,0 +1,95 @@
+const CustomDateTimePicker = (field) => {
+  const { initialData, onChange } = field;
+
+  const [date, setDate] = useState(initialData.date || getCurrentDate);
```
I guess it would be better to call the function immediately (`getCurrentDate`) instead of passing the reference for React to call it automatically to get the initial state value when `initialData.date` is not available:
`const [date, setDate] = useState(() => initialData.date || getCurrentDate());`

also include the lazy initialization `() =>`🔝,  like we did with time and period states initializations, so that the arrow function is only executed during the initial render, on subsequent re-renders, react knows it already has a state value and doesn't need to call the function again. The performance difference would be negligible here I believe, but it's better to use the arrow function for any initialization that creates objects, makes calculations or anything that could potentially be expensive.

asirvadAbrahamVarghese commented on app/javascript/components/date-time-picker/index.jsx:
```diff
@@ -0,0 +1,95 @@
+  const { initialData, onChange } = field;
+
+  const [date, setDate] = useState(initialData.date || getCurrentDate);
+  const [time, setTime] = useState(() => initialData.time || getCurrentTimeAndPeriod().time);
```
On initial render, time would look like this which is not in our expected "hh:mm" format, it should have been "09:54", But the invalid time error: "Enter a valid 12-hour time" is not displayed, probably thats because `isValid` state is initially true
<img width="994" height="156" alt="image" src="https://github.com/user-attachments/assets/1163e437-d2f7-4475-838e-6d71bba194e9" />
Right after time field is edited the error is displayed since our state `isValid` state is now updated:
<img width="1100" height="156" alt="image" src="https://github.com/user-attachments/assets/e8ee8a27-0bad-47f7-8419-22d46672dadb" />

**No issues with this implementation, this should work well** once we apply zero-padding to the hour value in `getCurrentTimeAndPeriod`(similar to what we did with minutes).
It should be padded after converting to 12-hour format, otherwise, the modulo will reduce it to a single digit again.
`hours = `${hours % 12 || 12}`.padStart(2, '0');`
Which should work well:
<img width="972" height="144" alt="image" src="https://github.com/user-attachments/assets/f13b969d-b32d-475f-9961-bd6817d76711" />

asirvadAbrahamVarghese commented on app/javascript/components/date-time-picker/index.jsx:
```diff
@@ -0,0 +1,95 @@
+  };
+
+  const handleDateChange = (newDate) => {
+    if (newDate.length > 0) {
```
`if (newDate.length)` should be fine here since 0 is also falsy in JS and any number greater than 0 will be truthy

asirvadAbrahamVarghese commented on app/javascript/components/date-time-picker/index.jsx:
```diff
@@ -0,0 +1,95 @@
+    const newTime = event.target.value;
+    setTime(newTime);
+    validateTime(newTime);
+    if (isValid) onChange({ value: combinedDateTime(), initialData });
```
Tried "11:5" in the time field, which doesn't match our pattern, so it's treated as an invalid time, the regex correctly returns false, which we also assign to the `isValid` state:
<img width="1896" height="902" alt="image" src="https://github.com/user-attachments/assets/970ef8bc-8cb8-4137-905e-76c8d69e0028" />
Given the condition(`if (isValid) onChange({ ... });`), we don't expect the `onChange` prop to be triggered in this case but it still gets called since `isValid` is still true(initial state value is true):
<img width="1738" height="864" alt="image" src="https://github.com/user-attachments/assets/10e57c65-8a08-4107-8215-5c5200538aa3" />
This is because state updates are asynchronous & batched and won't be reflecting immediately, when `setIsValid` is called react schedules the state update and updated value will be available only in the next render(only after the current event scope) not immediately after the call. So within the same event, we will still be accessing the previous state(which is `true`).

Returning the regex result from `validateTime` and using it for both setting state and controlling `onChange` should work fine:
```
  const validateTime = (value) => {
    const timeRegex = /^(0[1-9]|1[0-2]):[0-5][0-9]$/; // Matches 12-hour format hh:mm
    return timeRegex.test(value);
  };
  const handleTimeChange = (event) => {
    const newTime = event.target.value;
    setTime(newTime);
    const isValidTime = validateTime(newTime);
    setIsValid(isValidTime);
    if (isValidTime) onChange({ value: combinedDateTime(), initialData });
  };
```

asirvadAbrahamVarghese commented on app/javascript/components/date-time-picker/index.jsx:
```diff
@@ -0,0 +1,95 @@
+        year: 'numeric',
+      }).format(newDate[0]);
+      setDate(formattedDate);
+      onChange({ value: combinedDateTime(), initialData });
```
Like discussed [below](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2312990611), due to batched state updates we might want to update `combinedDateTime` to receive the updated date string:
```
  const combinedDateTime = ({ dateValue, timeValue, periodValue } = {}) =>
    `${dateValue || date} ${timeValue || time} ${periodValue || period}`;
```
and then from `handleDateChange` pass the date string value:
```
  const handleDateChange = (newDate) => {
    if (newDate.length) {
      const formattedDate = new Intl.DateTimeFormat('en-US', {
      .....
      onChange({
        value: combinedDateTime({ dateValue: formattedDate }),
        initialData,
      });
    }
  };
```

asirvadAbrahamVarghese commented on app/javascript/components/date-time-picker/index.jsx:
```diff
@@ -0,0 +1,95 @@
+
+  return (
+    <div>
+      <FormLabel>{field.label}</FormLabel>
```
**Is the label prop expected to always be provided? The label element still gets rendered in the DOM even when no label is provided:**
<img width="1526" height="1334" alt="image" src="https://github.com/user-attachments/assets/444c48a5-8d2e-4708-8d41-b723b92e007e" />
**conditionally rendering it can help prevent unused elements in the DOM:**
`{label && <FormLabel>{label}</FormLabel>}`
**So that Label won't be rendered:**
<img width="1710" height="1078" alt="image" src="https://github.com/user-attachments/assets/a64da1a7-1e9c-49ce-b104-eacd1efb3085" />

asirvadAbrahamVarghese commented on app/javascript/components/date-time-picker/index.jsx:
```diff
@@ -0,0 +1,95 @@
+  SelectItem,
+  FormLabel,
+} from 'carbon-components-react';
+import { getCurrentDate, getCurrentTimeAndPeriod } from '../service-dialog-form/helper';
```
Dealing with dates can be bug-prone. we should be fine for now with simple date logics but as logic grows, pulling in a utility like [date-fns](https://www.npmjs.com/package/date-fns) would help.

elsamaryv committed:
Service dialog edit form

elsamaryv committed:
updated

elsamaryv committed:
fixes issues with displaying values and saving of service dialog edit form

elsamaryv committed:
Updates cy test

elsamaryv committed:
Cypress tests fixed for service dialog edit form

elsamaryv committed:
Removes log statements and commented lines

elsamaryv committed:
updated prop name

elsamaryv committed:
Sets message as required

elsamaryv committed:
Updates on date-time picker components

elsamaryv committed:
updated

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/data.js:
```diff
@@ -0,0 +1,76 @@
+    skipErrors: [400],
+  }).then(() => {
+    // Redirect to the service dialogs explorer page after successful save
+    if (typeof onSuccess === 'function') {
```
Not sure about the best practice, but `if (onSuccess)` is good enough if we just want to guard against null or anything falsy

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-component-chooser.jsx:
```diff
@@ -0,0 +1,34 @@
+          className="component-item-wrapper"
+          draggable="true"
+          onDragStart={(event) => onDragStartComponent(event, dragItems.COMPONENT)}
+          key={index.toString()}
```
React uses key to track element identity and preserve state, using index as a key should be a last resort I think - [see doc](https://legacy.reactjs.org/docs/lists-and-keys.html)
Ideally, we should use a unique value from the data:
```
      list.map((item) => (
        <div
          key={`${item.dataKey}`}
```
Since we have `id` in the data, and if it’s unique, that can be used.

If we use index as the key and the list changes, say we delete or move a row, React might get confused about which item is which. That can lead to weird re-renders or the wrong DOM being reused.

Note: Index as keys should be fine if list won't change and items are purely presentational

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-field-actions.jsx:
```diff
@@ -0,0 +1,118 @@
+  const [{ showModal, ...editedFields }, setState] = useState({ showModal: false });
+
+  const onModalHide = () => setState((state) => ({ ...state, showModal: false }));
+  const onModalShow = () => setState((state) => ({ ...state, showModal: true }));
```
Just thinking if we could use single method instead of `onModalHide` & `onModalShow`
`const toggleModal = (show) => setState((state) => ({ ...state, showModal: show }));`
and use it like:
`() => toggleModal(false);`
`() => toggleModal(true);`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-field.jsx:
```diff
@@ -0,0 +1,80 @@
+
+  const fieldSelector = (fieldData) => {
+    const componentId = fieldData.field.componentId
+                       || (fieldData.field.type && getComponentIdFromType(fieldData.field.type))
```
This should work I guess:
```
    const componentId =
      fieldData.field.componentId ||
      getComponentIdFromType(fieldData.field.type || '');
```
When `fieldData.field.type` is falsy, switch will fall through to the default(`return 1`) when `''` is passed

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-field.jsx:
```diff
@@ -0,0 +1,80 @@
+  return (
+    <>
+      {
+        fieldSelector(fieldData)
```
Having the JSX directly in this spot instead of `fieldSelector` method shouldn't be a problem

elsamaryv committed:
Minor updates

elsamaryv committed:
Remove additional param

elsamaryv committed:
Replaces separate onModalHide and onModalShow fns with a single fn

elsamaryv committed:
updates fieldSelector method

elsamaryv committed:
Refactors DynamicField component

elsamaryv committed:
removes comment

elsamaryv committed:
Removes debugger

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx:
```diff
@@ -0,0 +1,120 @@
+  const inputId = `tab-${tabId}-section-${sectionId}-field-${fieldPosition}-checkbox`;
+  const editActionType = SD_ACTIONS.field.edit;
+
+  const refreshEnabledFields = section.fields
```
Looks good, but if they are larger arrays, better we drop the double iteration and use [reduce](https://www.geeksforgeeks.org/javascript/javascript-array-reduce-method/), which can handle it in one shot:
```
  const refreshEnabledFields = section.fields.reduce((result, field) => {
    if (field.showRefresh) {
      result.push({ value: field.label, label: field.label });
    }
    return result;
  }, []);
```

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx:
```diff
@@ -0,0 +1,120 @@
+    setFieldState((prevState) => ({ ...prevState, dynamic: isDynamic }));
+  };
+
+  const ordinaryCheckboxOptions = () => ([
```
Can these be regular constants instead of arrow functions?
```
  const ordinaryCheckboxOptions = [
    dynamicFields.defaultCheckboxValue,
    dynamicFields.required,
    dynamicFields.readOnly,
    dynamicFields.visible,
    dynamicFields.fieldsToRefresh,
  ];
  const dynamicCheckboxOptions = [
    dynamicFields.automateEntryPoint,
    dynamicFields.showRefresh,
    dynamicFields.loadOnInit,
    dynamicFields.required,
    dynamicFields.fieldsToRefresh,
  ];
```
From what I understand, whether it’s a const or a function, it gets recreated on every re-render.
If these values are fully static, we can even move it outside the component, that way there’s only one copy and it won’t be recreated on every render(whether its a function or a const):
```
  const ordinaryCheckboxOptions = [
    dynamicFields.defaultCheckboxValue,
    dynamicFields.required,
    dynamicFields.readOnly,
    dynamicFields.visible,
    dynamicFields.fieldsToRefresh,
  ];

  const dynamicCheckboxOptions = [
    dynamicFields.automateEntryPoint,
    dynamicFields.showRefresh,
    dynamicFields.loadOnInit,
    dynamicFields.required,
    dynamicFields.fieldsToRefresh,
  ];

/** Component to render a Field. */
const DynamicCheckbox = .......
```

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-date-picker.jsx:
```diff
@@ -0,0 +1,130 @@
+
+/** Component to render a Field. */
+const DynamicDatePicker = ({ dynamicFieldData: { section, field, fieldPosition }, onFieldAction }) => {
+  const { tabId, sectionId } = section;
```
Same thoughts as I shared([here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2454339788)) on `app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-date-picker.jsx:
```diff
@@ -0,0 +1,130 @@
+    setFieldState((prevState) => ({ ...prevState, dynamic: isDynamic }));
+  };
+
+  const ordinaryDatePickerOptions = () => ([
```
Same thoughts as I shared([here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2454341510)) on `app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-dropdown.jsx:
```diff
@@ -0,0 +1,178 @@
+
+/** Component to render a Field. */
+const DynamicDropdown = ({ dynamicFieldData: { section, field, fieldPosition }, onFieldAction }) => {
+  const { tabId, sectionId } = section;
```
Same thoughts as I shared([here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2454339788)) on `app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-dropdown.jsx:
```diff
@@ -0,0 +1,178 @@
+  const inputId = `tab-${tabId}-section-${sectionId}-field-${fieldPosition}-dropdown`;
+  const editActionType = SD_ACTIONS.field.edit;
+
+  const refreshEnabledFields = section.fields
```
Same thoughts as I shared([here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2454340717)) on `app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx`
This seems like redundant...

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-dropdown.jsx:
```diff
@@ -0,0 +1,178 @@
+    setFieldState((prevState) => ({ ...prevState, automationType: value }));
+  };
+
+  const ordinaryDropdownOptions = () => ([
```
Same thoughts as I shared([here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2454341510)) on `app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-dropdown.jsx:
```diff
@@ -0,0 +1,178 @@
+        dynamicFieldAction={(event, inputProps) => fieldActions(event, inputProps)}
+        fieldConfiguration={DropdownEditFields()}
+        dynamicToggleAction={(isDynamic) => resetEditModalTabs(isDynamic)}
+        onValueChange={(value) => updateAutomationType(value)}
```
Same thoughts as I shared([here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2454343399)) on `app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-radio-button.jsx:
```diff
@@ -0,0 +1,155 @@
+        updateFieldProps={handleFieldUpdate}
+        dynamicFieldAction={(event, inputProps) => fieldActions(event, inputProps)}
+        fieldConfiguration={radioButtonEditFields()}
+        dynamicToggleAction={(isDynamic) => resetEditModalTabs(isDynamic)}
```
Same thoughts as I shared([here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2454343399)) on `app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-tag-control.jsx:
```diff
@@ -0,0 +1,186 @@
+        }));
+      });
+    }
+  }, [fieldState.categories.length]);
```
I hope we’re setting the categories only on the initial render. If so, we can pass an empty array to useEffect instead of length. Is ESLint complaining about that?

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-tag-control.jsx:
```diff
@@ -0,0 +1,186 @@
+    return tabs;
+  };
+
+  const sortedItems = () => {
```
Same thoughts as I shared([here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2454400025)) on `app/javascript/components/service-dialog-form/dynamic-fields/dynamic-radio-button.jsx`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-text-area.jsx:
```diff
@@ -0,0 +1,128 @@
+
+  const inputId = `tab-${tabId}-section-${sectionId}-field-${fieldPosition}-text-area`;
+  const editActionType = SD_ACTIONS.field.edit;
+  const refreshEnabledFields = section.fields
```
Same thoughts as I shared([here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2454340717)) on `app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-text-area.jsx:
```diff
@@ -0,0 +1,128 @@
+        updateFieldProps={handleFieldUpdate}
+        dynamicFieldAction={(event, inputProps) => fieldActions(event, inputProps)}
+        fieldConfiguration={textAreaEditFields()}
+        dynamicToggleAction={(isDynamic) => resetEditModalTabs(isDynamic)}
```
Same thoughts as I shared([here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2454343399)) on `app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-text-input.jsx:
```diff
@@ -0,0 +1,130 @@
+          placeholder={__('Default value')}
+          value={fieldState.value}
+          readOnly={fieldState.readOnly}
+          onChange={(e) => handleFieldUpdate(e, { value: e.target.value })}
```
Same thoughts as I shared([here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2454342395)) on `app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-text-input.jsx:
```diff
@@ -0,0 +1,130 @@
+        updateFieldProps={handleFieldUpdate}
+        dynamicFieldAction={(event, inputProps) => fieldActions(event, inputProps)}
+        fieldConfiguration={textBoxEditFields()}
+        dynamicToggleAction={(isDynamic) => resetEditModalTabs(isDynamic)}
```
Same thoughts as I shared([here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2454343399)) on `app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-time-picker.jsx:
```diff
@@ -0,0 +1,249 @@
+
+/** Component to render a Field. */
+const DynamicTimePicker = ({ dynamicFieldData: { section, field, fieldPosition }, onFieldAction }) => {
+  const { tabId, sectionId } = section;
```
Same thoughts as I shared([here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2454339788)) on `app/javascript/components/service-dialog-form/dynamic-fields/dynamic-checkbox.jsx`

elsamaryv committed:
checkbox refactoring

elsamaryv committed:
refactors checkbox

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/dynamic-section.jsx:
```diff
@@ -0,0 +1,144 @@
+      tab={tabId}
+      section={sectionId}
+      id={`dynamic-tab-${tabId}-section-${sectionId.toString()}`}
+      key={`dynamic-tab-${tabId}-section-${sectionId.toString()}`}
```
I just noticed there’s an issue with how `sectionId` is being used in the key here
(just pointing to the id value here, since it’s the same one we use for both id and key)
Steps to reproduce:
1) Setting up 2 sections with some fields here:
<img width="2524" height="1562" alt="image" src="https://github.com/user-attachments/assets/a324ff0f-138d-4dac-8d57-84df9c459e22" />
2) Removing the first one:
<img width="1484" height="1576" alt="image" src="https://github.com/user-attachments/assets/23f9c17b-26ba-4271-8fe6-a8398bb8c105" />
3) Now adding another section and looking into the id value we can see that both are same
<img width="2766" height="1602" alt="image" src="https://github.com/user-attachments/assets/5a3b2a7b-2165-4cb2-b000-74e2af0068ef" />
4) Now try to add a field to the second section
<img width="2438" height="1256" alt="image" src="https://github.com/user-attachments/assets/9eaf94e5-e8a1-4815-974d-dd87925179b8" />
5) But that ends up on to the first section:
<img width="2664" height="1544" alt="image" src="https://github.com/user-attachments/assets/c29e869f-0eb2-4226-b951-2e6d8401b4fb" />

Just like we discussed in [this](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592#discussion_r2439271837) comment about react keys, this was also because react got confused since two elements have the same key and incorrectly updated the wrong component.
I am guessing, problem could be [here](https://github.com/elsamaryv/manageiq-ui-classic/blob/service-dialogs-conversion/app/javascript/components/service-dialog-form/index.jsx#L292) where we are using `sections.length` as `sectionId`

elsamaryv committed:
datepicker - initial refactoring

elsamaryv committed:
Moves getRefreshEnabledFields fn to helper class

elsamaryv committed:
refactors text area and text input fields

elsamaryv committed:
refactors time-picker component

elsamaryv committed:
Adds reset  button to edit form

elsamaryv committed:
Disables reset button at first

elsamaryv committed:
Modifies the reset button's visibility based on the form

elsamaryv committed:
Modifies save button visibility based on form

elsamaryv committed:
Fix - Edit modal remains draggable in open state

elsamaryv committed:
Implements a unique ID counter for sections to prevent id mismatch

elsamaryv committed:
Removes duplication

elsamaryv committed:
Disables unwanted interactions on edit field modal

To prevent drag event when edit modal is open

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/edit-field-modal/index.jsx:
```diff
@@ -0,0 +1,200 @@
+import InlineFlashMessage from '../../common/inline-flash-message';
+
+const EditFieldModal = ({
+  componentId, fieldConfiguration, showModal, onModalHide, onModalApply, initialData,
```
I think `onModalApply` & `showModal` are unused here, may be we can consider removing it

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/edit-field-modal/index.jsx:
```diff
@@ -0,0 +1,200 @@
+
+    if (data.initialData.label === 'Timepicker') {
+      onTimePickerChange(data.value);
+    } else if (data.initialData.label === 'Entry point') {
```
1. It looks like the Select Entry Point modal isn’t rendering correctly, half of it shows inside the parent and the rest spills to the right with a scrollbar:
<img width="2974" height="1702" alt="image" src="https://github.com/user-attachments/assets/8282d3ba-2d37-4611-bf28-f13336b4e176" />
Can we make the Edit Field modal wider and possibly reduce the size of the Select Entry Point modal to fit inside? I was thinking of something like a size prop, similar to react-bootstrap’s Modal, but it doesn’t look like Carbon offers that.

Seems like this modal also flickers when the mouse leaves the modal area as I mentioned [here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592#discussion_r2588695938)


2. Should the ‘Remove Entry Point’ button be disabled when nothing is selected?
<img width="1504" height="600" alt="image" src="https://github.com/user-attachments/assets/e0d8dd44-92af-46b9-b3d0-aa527cf75eb4" />

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/edit-tab-modal/index.jsx:
```diff
@@ -0,0 +1,48 @@
+        schema={createSchema(usedTabNames, tabInfo.name)}
+        initialValues={{
+          tab_name: tabInfo.name,
+          tab_description: tabInfo.description,
```
Looks like the Edit Tab modal isn’t storing the description field value after saving and reopening:
<img width="1938" height="1084" alt="image" src="https://github.com/user-attachments/assets/4d0258f2-fba8-4007-bb84-c8bbdba7a5cb" />
<img width="1838" height="972" alt="image" src="https://github.com/user-attachments/assets/1b3408c8-087d-4571-8fe6-65fd3623d816" />

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/edit-field-modal/index.jsx:
```diff
@@ -0,0 +1,200 @@
+  };
+
+  // Custom form template
+  const FormTemplate = ({ formFields }) => {
```
Better if we can move this `FormTemplate` component outside of `EditFieldModal` component
It gets recreated on every render of EditFieldModal

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/index.jsx:
```diff
@@ -0,0 +1,730 @@
+  const dragEnterItem = useRef(null); /** Stores the information of component where the dragged item is being hovered before release. */
+  const draggedItem = useRef(null); /** Stores the information of component being dragged. */
+  const hoverItem = useRef(null); /** Stores the tab and section position during the drop event. */
+  const nextSectionId = useRef(1); /** Counter for generating unique section IDs */
```
To me it looks neat to use `useRef` for this, since the counter doesn’t need to cause unnecessary re-renders and the value sticks around
I’m just not sure if there are any downsides compared to using the state managing hooks like `useState`/`useReducer`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/index.jsx:
```diff
@@ -0,0 +1,730 @@
+    evaluateSubmitButton();
+  }, [data]);
+
+  const onTabAction = (type, event) => tabAction({
```
Looks like `onTabAction` is not used

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/index.jsx:
```diff
@@ -0,0 +1,730 @@
+      description: data.description,
+      dialog_tabs: data.formFields
+        .filter((tab) => tab.tabId !== 'new') // Filter out the "Create new tab" option
+        .map((tab, index) => ({
```
Maybe we could turn these `filter` and `map` iterations into a single `reduce` to avoid looping over the array multiple times?

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/index.jsx:
```diff
@@ -0,0 +1,730 @@
+          }
+        </div>
+        {/* Form submit/cancel buttons */}
+        <div className="custom-button-wrapper">
```
Same thought as I mentioned [here](https://github.com/ManageIQ/manageiq-ui-classic/pull/9592/files#r2591993385) about class names

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/helper.js:
```diff
@@ -0,0 +1,544 @@
+  return result;
+};
+
+const getFieldsInfo = (fields) => {
```
If there is no other logic to add, we can use `parseFieldsInfo` directly when required instead of a wrapper `getFieldsInfo`

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/helper.js:
```diff
@@ -0,0 +1,544 @@
+});
+
+export const formattedCatalogPayload = (data) => {
+  const payload = payloadForSave(data);
```
We can do this directly instead of using `payloadForSave`
```
export const formattedCatalogPayload = (data) => {
  const payload = { action: 'create', resource: getDialogInfo(data) };
  return payload;
};
```

asirvadAbrahamVarghese commented on app/javascript/components/service-dialog-form/helper.js:
```diff
@@ -0,0 +1,544 @@
+  const trimmed = value.trim().toLowerCase();
+
+  const isDuplicate = usedNames
+    .filter((name) => name.toLowerCase() !== currentName.toLowerCase())
```
I think we can do this using `some` instead of `filter`, `map` and `includes`
```
const isDuplicate = usedNames.some((name) => 
  name.toLowerCase() !== currentName.toLowerCase() && 
  name.toLowerCase() === trimmed
);
```

asirvadAbrahamVarghese commented on app/javascript/packs/component-definitions-common.js:
```diff
@@ -262,12 +265,13 @@ ManageIQ.component.addReact('MiqAlertSetForm', MiqAlertSetForm);
 ManageIQ.component.addReact('MiqCustomTab', MiqCustomTab);
 ManageIQ.component.addReact('MiqDataTable', MiqDataTable);
 ManageIQ.component.addReact('MiqMarkdown', MiqMarkdown);
+ManageIQ.component.addReact('ServiceDialogForm', ServiceDialogForm);
```
Do we need to maintain alphabetical order here?

jrafanie commented on cypress/e2e/ui/Automation/Embedded-Automate/customization/service-dialogs/add/checkbox.cy.js:
```diff
@@ -0,0 +1,411 @@
+    cy.dragAndDropComponent('Check Box');
+  });
+
+  it('should verify default properties of the checkbox', () => {
```
The refactoring doesn't need to happen right now, but we're more likely to do it now vs. later.

jrafanie commented on cypress/e2e/ui/Automation/Embedded-Automate/customization/service-dialogs/add/checkbox.cy.js:
```diff
@@ -0,0 +1,411 @@
+
+    // Enable dynamic option
+    cy.get('.edit-field-modal input[name="dynamic"]')
+      .check({ force: true });
```
`force: true` statements should have an explanation why they're needed... if it's being covered by another element, maybe there's a better way to do it by waiting for the object to be fully visible and accessible.

elsamaryv committed:
Make tabs draggable

elsamaryv committed:
Identify tabs by unique id instead of array position

elsamaryv committed:
Uses moment js to format date

elsamaryv committed:
This commit will be reverted later (TO DO)

elsamaryv committed:
Refactorings

elsamaryv committed:
Refactors handleSubmit()

elsamaryv committed:
uses reduce instead of filter and map

elsamaryv committed:
Fixes tab reordering conflict after saving

elsamaryv committed:
Refactored dynamic-field.jsx & removed the unnecessary fieldSelector function wrapper.

elsamaryv committed:
Refactors dropdown component

elsamaryv commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-dropdown.jsx:
```diff
@@ -0,0 +1,178 @@
+          helperText={__('This is helper text')}
+          items={sortedItems()}
+          sortItems={(items) => items}
+          itemToString={(item) => (item ? item.description : '')}
```
These are required. Without these, Carbon would re-sort our items and display [object Object] instead of readable text.

elsamaryv committed:
Refactors radio button component

elsamaryv committed:
Moves sortIems function to helper to remove redundancy

elsamaryv commented on app/javascript/components/service-dialog-form/dynamic-fields/dynamic-time-picker.jsx:
```diff
@@ -0,0 +1,249 @@
+        <DatePicker
+          datePickerType="single"
+          onChange={handleDateChange}
+          minDate={fieldState.showPastDates ? undefined : new Date().toLocaleDateString()}
```
Same as https://github.com/ManageIQ/manageiq-ui-classic/pull/9592#discussion_r2469346534

elsamaryv committed:
Refactors tag control component

elsamaryv committed:
Deleting a tab should move the focus to the tab on the left, not the right

elsamaryv committed:
Fix - Saving a default value for one field deletes this value for other fields

elsamaryv committed:
Adds error handling during form submit

elsamaryv committed:
suppress default api error

elsamaryv committed:
Reuse existing InlineFlashMessage component

elsamaryv committed:
Adds + icon to "create tab"

elsamaryv committed:
updated

elsamaryv committed:
Fix - modal no longer hides when mouse is moved out of the field area

The drag-and-drop cursor icon no longer appears when hovering over the Edit Field modal.
Draggable elements are disabled while the modal is open, preventing accidental drag operations.
The modal remains fully interactive and visible until explicitly closed by the user.

elsamaryv committed:
Removed unused props and functions for cleaner code

elsamaryv committed:
Use onModalHide directly instead of wrapping it with a separate onCancel

elsamaryv committed:
Fixes issue with saving the tab description

elsamaryv committed:
updated

elsamaryv committed:
Removes unused function

elsamaryv committed:
Improved code organization with all state declarations at the top of the component

elsamaryv committed:
Update index.jsx

Removed unused functions

elsamaryv committed:
Directly call parseFieldsInfo in getSectionsInfo

elsamaryv commented:
> @elsamaryv This is a good start so far but lots of issues that still need to be fixed. We can tackle them a bit at a time to make it easier. To start here are the issues I found so far:
> 
> **Overall**
> 
> * Not needed but if possible try adding a Reset Button to the Edit form
> * Form submit error handling, show error
> * Modal keeps closing when you hover away from the page
> * Deleting a tab should move the focus to the tab on the left, not the right. This avoids the focus moving to the Create Tab tab which is empty
> * Saving a default value for 1 field deletes this value for other fields, not sure why
> * DO NOT FIX THIS ONE YET: we can leave this for the final review. This can be avoided possibly and use random number, need to ask Jason first if random number is ok. Need to try to make Default Field name to be formatted like text_box_1, text_box_2. Create array of all field names for each field type and try to create text_box_1. If text_box_1 exists then name is text_box_2. If text_box_2 exists try text_box_3 and repeat until a valid name is found.
> * Common modal issues:
>   
>   * Help field doesn’t save
>   * Validation fields don’t save
>   * Validator field missing tooltip saying Regular Expression
> 
> **Tabs**
> 
> * Try to add carbon icon to the Create Tab label to make it look like the old form
> * Issue with saving the tab description
> 
> **Textbox**
> 
> * Value type field doesn’t save
> 
> **Text Area**
> 
> * Dialog Editor form can’t save when this field is required and is empty

TO DO (from the cases mentioned above):

- Common modal issues:

Help field doesn’t save
Validation fields don’t save
Validator field missing tooltip saying Regular Expression

- Textbox - Value type field doesn’t save

- Text Area - Dialog Editor form can’t save when this field is required and is empty

elsamaryv committed:
commented lines removed

elsamaryv committed:
Updated files to use @carbon/react instead of the deprecated carbon-components-react package

This fixes the webpack build error

elsamaryv committed:
Replaces '@carbon/icons-react' with '@carbon/react/icons' for carbon 11

elsamaryv committed:
updated

elsamaryv committed:
Textarea layout fixes

elsamaryv committed:
Fix tab layout issues after Carbon upgrade

elsamaryv committed:
Fix helper field(text areas) inside the edit modal from expanding too much

elsamaryv committed:
Update default value field for text area component in the edit modal

elsamaryv committed:
Restrict textarea vertical expansion in dynamic form fields

elsamaryv committed:
Fix radio button rendering and standardize data structure for both radio button and dropdown components

elsamaryv committed:
Fix dropdown initial values display issue

elsamaryv committed:
Fix Carbon MultiSelect dropdown styling to remove bullets on items

elsamaryv committed:
Fix Carbon MultiSelect dropdown styling in modal

elsamaryv committed:
Fix layout issue with the Field array component for adding entries

The column now appears side-by-side and aligned

elsamaryv committed:
Update class

elsamaryv committed:
Remove default legend border/underline
